### ISSUE TITLE
Cookie auth, client redirect on expiration

### ISSUE THREAD

**Issue by los93sol** [aspnet/AspNetCore#3888]
I have been using built in cookies auth since the previews and it has been great to date.  Recently I thought it would be nice to implement a feature that prompt the user a few minutes before their session will expire and assumed it would be a trivial task.  It turned out to be anything but that and the only approaches I could come up with were the following...

1. Hack CookieManager to add a clear text cookie with nothing more than the expiration so I can get to it from the client via js 
2. Hack an instance of DataProtector from a controller method and access it and decrypt it from the client like an API
3. Use context.AuthorizeAsync to get at the ticket as an API

I was able to get this working but it feels sloppy as the entire way AuthBuilders work ends up newing all of their dependencies so you can’t just DI the component you need to extend functionality like this easily.  Additionally I think I ran into a bug where when implementing an ITicketStore the CookieAuthHandler retrieves the current ticket, then invokes various middleware and matched routes and only after that is the ticket renewed.  This is problematic because when I access the ticket in middleware or an api or controller the value I get is potentially stale and different from what is actually in the current ticket so I had to make a second request to make sure it is in sync.

So my questions are....
1. Is this a bug?
2. Is there a better way to do this?
3. Any consideration to letting the framework inject some script into the client to do most of this for us?
4. Was this a consideration during design or am I missing something?  It seems like redirecting the client on expiration would be a common use case instead of having the user clicking about wondering why all those buttons they click firing requests to APIs and hubs aren’t doing anything.

**Comment by los93sol** [comment]
@tratcher While I agree with the sentiment of API responses an acting on them, it’s a cleaner presentation to pre-emptively redirect the user to a page that clearly tells them they were logged out due to inactivity.  In my case I have a lot of monitoring type tools that my users will load up, slide it over to a second monitor and just glance over throughout their day.  In this scenario I wanted to be able to pop a simple modal “You’re session will expire at X” and a butting to extend the session.  When they extend the session it just makes some noise to an endpoint and grabs the new expiration and sets a new timer on the client.  If they didn’t click it before expiration then the redirect kicks in.  It works now with the caveat that getting at the expiration is a bit hacky and when I extend the session I have to make two calls to the server because the first one is stale due to RenewAsync not being called until after the pipeline and any chance to write to the response body.  Really just a hook to get the ticket without having to call AuthorizeAsync and the renewal happening earlier in the pipeline would ha e made this quite a bit easier.  In the end it was simple but I spent a lot of time digging through framework code to sort out how to get at the ticket and then figuring out why the first response was always stale

**Comment by Tratcher** [comment]
Yeah, there's no first class way to track that from the client side right now. Want to share a code sample of how you got it working so we can consider it for future design improvements?

**Comment by los93sol** [comment]
Absolutely, I’ll try to get a barebones implementation up with what I did to get this working in the next few days.  Thanks for the quick responses!

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
